src/services/refactors/convertToOptionalChainExpression.ts

```diff
@@ -0,0 +1,126 @@
+        const expression = containingNode && isExpressionStatement(containingNode) ? containingNode.expression : containingNode;
+        if (!expression) return undefined;
+
+        const checker = program.getTypeChecker();
+
+
+        if (isBinaryExpression(expression)) {
+            const start = forEmptySpan ? expression.pos : startToken.pos;
+
+            const fullPropertyAccess = getFullPropertyAccessChain(expression);
+            if (!fullPropertyAccess) return undefined;
+            if (expression.operatorToken.kind !== SyntaxKind.AmpersandAmpersandToken && expression.operatorToken.pos <= fullPropertyAccess.pos) return undefined;
+
+            // ensure that each sequential operand in range matches the longest acceess chain
+            let checkNode = expression.left;
+            let firstOccurrence: PropertyAccessExpression | Identifier = fullPropertyAccess;
+            while (isBinaryExpression(checkNode) && checkNode.operatorToken.kind === SyntaxKind.AmpersandAmpersandToken && (isPropertyAccessExpression(checkNode.right) || isIdentifier(checkNode.right)) && checkNode.right.pos >= start) {
+                if (!checker.containsMatchingReference(fullPropertyAccess, checkNode.right)) {
+                    return undefined;
+                }
+                firstOccurrence = checkNode.right;
+                checkNode = checkNode.left;
+            }
+            // check final identifier
+            if ((isIdentifier(checkNode) || isPropertyAccessExpression(checkNode)) && checker.containsMatchingReference(fullPropertyAccess, checkNode) && checkNode.pos >= start) {
+                firstOccurrence = checkNode;
+            }
+            return firstOccurrence ? { fullPropertyAccess, firstOccurrence, expression } : undefined;
+        }
+
+        if (isConditionalExpression(expression)) {
+            const whenTrue = expression.whenTrue;
+            const condition = expression.condition;
+
+            if ((isIdentifier(condition) || isPropertyAccessExpression(condition)) && isPropertyAccessExpression(whenTrue) && checker.containsMatchingReference(whenTrue, condition)) {
+                return { fullPropertyAccess: whenTrue, firstOccurrence: condition, expression };
+            }
+        }
+        return undefined;
+    }
+
+    function getRightHandSidePropertyAccess(node: BinaryExpression | CallExpression): PropertyAccessExpression | undefined {
+        if (isCallExpression(node) && isPropertyAccessExpression(node.expression)) {
+            // a && |a.b|();
+            return node.expression;
+        }
+        else if (isBinaryExpression(node)) {
+            if (isPropertyAccessExpression(node.left)) {
+                // a && |a.b| == 1;
+                return node.left;
+            }
+            else if (isCallExpression(node.left) && isPropertyAccessExpression(node.left.expression)) {

```

Comment:
this feels like a missed opportunity for recursion, but

1. it might not make sense at all here
2. it might only apply in weird cases
3. it might make the rewrite pretty complex

did you consider recurring on this line instead of manually checking just the CallExpr case?

Reply:
The smart thing to do is probably to recurse to find the start and end nodes of a valid `&&` chain. It seemed like there were a few simple cases for what could happen on the RHS, but that may not be the case for more complicated expressions.

Reply:
from discussing it together: recursion on the call is technically possible but it would be REALLY WEIRD to write `a && a.b()()() === 1`

Reply:
What about `a && a.b().c` though?

Reply:
Also a.b()()()?

Reply:
`a && 1 == a.b` was also discussed.

Reply:
I added some tests for the not supported cases for now, but I am working on getting them in.